ray-operator/controllers/ray/expectations/scale_expectations.go

```diff
@@ -0,0 +1,178 @@
+	Delete(rayClusterName, namespace string)
+}
+
+func NewRayClusterScaleExpectation(client client.Client) RayClusterScaleExpectation {
+	return &realRayClusterScaleExpectation{
+		Client:     client,
+		itemsCache: cache.NewIndexer(rayPodKey, cache.Indexers{GroupIndex: groupIndexFunc, RayClusterIndex: rayClusterIndexFunc}),
+	}
+}
+
+type realRayClusterScaleExpectation struct {
+	client.Client
+	// itemsCache is only used to cache rayPod.
+	itemsCache cache.Indexer
+}
+
+func (r *realRayClusterScaleExpectation) ExpectScalePod(namespace, rayClusterName, group, name string, action ScaleAction) {
+	// Strictly limit the data type stored in itemsCache to rayPod.
+	// If an error occurs, it must be due to an issue with our usage. We should panic immediately instead of returning an error.
+	if err := r.itemsCache.Add(&rayPod{
+		name:            name,
+		namespace:       namespace,
+		group:           group,
+		rayCluster:      rayClusterName,
+		action:          action,
+		recordTimestamp: time.Now(),
+	}); err != nil {
+		// If an error occurs, it indicates that there is an issue with our KeyFunc.
+		// This is a fatal error, panic it.
+		panic(err)
+	}
+}
+
+func (r *realRayClusterScaleExpectation) IsSatisfied(ctx context.Context, namespace, rayClusterName, group string) (isSatisfied bool) {
+	items, err := r.itemsCache.ByIndex(GroupIndex, fmt.Sprintf("%s/%s/%s", namespace, rayClusterName, group))
+	if err != nil {
+		// An error occurs when there is no corresponding IndexFunc for GroupIndex. This should be a fatal error.
+		panic(err)
+	}
+	isSatisfied = true
+	for i := range items {
+		rp := items[i].(*rayPod)
+		pod := &corev1.Pod{}
+		isPodSatisfied := false
+		switch rp.action {
+		case Create:
+			if err := r.Get(ctx, types.NamespacedName{Name: rp.name, Namespace: namespace}, pod); err == nil {
+				isPodSatisfied = true
+			} else {
+				// Tolerating extreme case:
+				//   The first reconciliation created a Pod. If the Pod was quickly deleted from etcd by another component
+				//   before the second reconciliation. This would lead to never satisfying the expected condition.
+				//   Avoid this by setting a timeout.
+				isPodSatisfied = errors.IsNotFound(err) && rp.recordTimestamp.Add(ExpectationsTimeout).Before(time.Now())
+			}
+		case Delete:
+			if err := r.Get(ctx, types.NamespacedName{Name: rp.name, Namespace: namespace}, pod); err != nil {
+				isPodSatisfied = errors.IsNotFound(err)
+			} else {
+				isPodSatisfied = pod.DeletionTimestamp != nil
+			}
+		}
+		// delete satisfied item in cache
+		if isPodSatisfied {
+			if err := r.itemsCache.Delete(items[i]); err != nil {
+				// Fatal error in KeyFunc.
+				panic(err)
+			}
+		} else {
+			isSatisfied = false
+		}
+	}
+	return isSatisfied
+}
+
+func (r *realRayClusterScaleExpectation) Delete(rayClusterName, namespace string) {
+	items, err := r.itemsCache.ByIndex(RayClusterIndex, fmt.Sprintf("%s/%s", namespace, rayClusterName))
```

Comment:
What will happen if there is no key `fmt.Sprintf("%s/%s", namespace, rayClusterName)` in the indexer? Does it only return an empty list?